### ISSUE TITLE
Clarify what should go in public- and private namespaces of cops

### ISSUE THREAD

**Issue by Drenmi** [bbatsov/rubocop#4439]
Currently, there are a few different takes on what should go in the `private` namespace of cops. Some cops only expose their callbacks, some expose `#autocorrect`, and some frivolously have entirely `public` interfaces.

Technically, the only methods required to go in the `public` namespace are the callbacks. Even `#autocorrect` works fine as `private`. (Constants naturally go at the top, since they aren't affected by visibility declarations anyway.)

Are there any compelling reasons to make methods other than the callbacks `public`? Ping @jonas054, @rrosenblum, @bbatsov.

We now have an `InternalAffairs` department, so we could actually enforce this programmatically if we wanted.

**Comment by backus** [comment]
The specs pass (with one other tweak that I don't think is important) if I add this to the end of `lib/rubocop.rb`:

```ruby
allowed = [
  :join_force?,
  :investigate,
  :before_entering_scope,
  :after_entering_scope,
  :before_leaving_scope,
  :after_leaving_scope,
  :before_declaring_variable,
  :after_declaring_variable,
  :check,
  :message,
  :validate_config
]

RuboCop::Cop::Cop.all.each do |cop|
  cop.public_instance_methods(false).grep_v(/\Aon_/).each do |method|
    next if allowed.include?(method)

    cop.class_eval { private(method) }
  end
end
```

So it looks like these methods 

- `join_force?`
- `investigate`
- `before_entering_scope`
- `after_entering_scope`
- `before_leaving_scope`
- `after_leaving_scope`
- `before_declaring_variable`
- `after_declaring_variable`
- `check`
- `message`
- `validate_config`

plus the `on_` callbacks are the public API.

**Comment by Drenmi** [comment]
Nice one, @backus! I wonder if `#check` and `#message` have special cases in a few cops, because I have several cops working where they are `private`. Maybe some specs are testing these directly?

**Comment by backus** [comment]
@Drenmi yeah I think they are being called directly

**Comment by jonas054** [comment]
:+1: It's OK with me if you make everything private that can be private.

**Comment by Drenmi** [comment]
> It's OK with me if you make everything private that can be private.

👍
